src/libraries/Hooks.sol

```diff
@@ -65,42 +83,68 @@ library Hooks {
                 || permissions.afterSwap != self.hasPermission(AFTER_SWAP_FLAG)
                 || permissions.beforeDonate != self.hasPermission(BEFORE_DONATE_FLAG)
                 || permissions.afterDonate != self.hasPermission(AFTER_DONATE_FLAG)
+                || permissions.beforeSwapReturnDelta != self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+                || permissions.afterSwapReturnDelta != self.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG)
+                || permissions.afterAddLiquidityReturnDelta != self.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                || permissions.afterRemoveLiquidityReturnDelta
+                    != self.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
         ) {
             revert HookAddressNotValid(address(self));
         }
     }
 
     /// @notice Ensures that the hook address includes at least one hook flag or dynamic fees, or is the 0 address
     /// @param self The hook to verify
+    /// @return bool True if the hook address is valid
     function isValidHookAddress(IHooks self, uint24 fee) internal pure returns (bool) {
+        // The hook can only have a flag to return a hook delta on an action if it also has the corresponding action flag
+        if (
```

Comment:
nit: would find it much more readable if these were separate ifs for each failure case

```
if (!BEFORE_SWAP && BEFORE_SWAP_RETURNS_DELTA) return false;
if (!...) return false
...
```